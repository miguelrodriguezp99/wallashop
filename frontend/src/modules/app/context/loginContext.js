import { createContext, useState } from "react";

// Este es el que tenemos que consumir
export const LogginContext = createContext();

// Este es el que nos provee de acceso al contexto
export function LogginProvider({ children }) {
	const [token, setToken] = useState(null);
	const [image, setImage] = useState(null);
	const [user, setUser] = useState({});

	return (
		<LogginContext.Provider
			value={{
				token,
				setToken,
				image,
				setImage,
				user,
				setUser,
			}}
		>
			{children}
		</LogginContext.Provider>
	);
}
