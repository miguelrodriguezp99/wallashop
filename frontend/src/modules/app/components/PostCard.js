import { Button } from "@nextui-org/react";
import React, { useContext } from "react";
import { RiDeleteBin7Line } from "react-icons/ri";
import { useNavigate } from "react-router-dom";
import { toast } from "sonner";
import { deletePost } from "../../../backend/postService";
import { LogginContext } from "../context/loginContext.js";
import EditPostModal from "./modals/EditPostModal.js";
import StompService from "./../socket/StompService.js";
import { config } from "../../../config/constants.js";

import "../styles/posts.css";

const PostCard = ({ post }) => {
	const { user } = useContext(LogginContext);
	const navigate = useNavigate();

	if (post.stillValid != null) {
		var date = post.stillValid.split("T").slice(0, 1);
		var time = post.stillValid
			.split("T")
			.slice(1, 2)
			.join(" ")
			.split(".")
			.slice(0, 1);
	}

	const handleDelete = () => {
		deletePost(post.id, onSuccess, onErrors);
	};

	const onSuccess = () => {
		toast.success("Post borrado correctamente.");

		const bearer = localStorage.getItem(config.SERVICE_TOKEN_NAME);
		if (StompService.client && StompService.client.connected) {
			StompService.send("/app/handleNotifications", JSON.stringify({success: true, userId: bearer.slice(-7) }));
			
		} 

		navigate("/wallashop/");
		window.location.reload(true);
	};

	const onErrors = () => {
		toast.error("Error al borrar el post.");
	};
	return (
		<>
			<article className="p-5 transform duration-300 hover:-translate-y-1 cursor-pointer group hover:shadow-white bg-cardback ml-2 rounded-md">
				{post.type === "OFFER" ? (
					<div className="folded-ribbon text-lg bg-pink-400">Oferta</div>
				) : (
					<div className="folded-ribbon text-lg bg-purple-500">Cupón</div>
				)}
				<div className=" relative max-h-125 overflow-hidden">
					<img className="absolute rounded-md" src={post.images[0]} alt="" />
					{post.images.length > 1 ? (
						<img
							className="relative transform duration-500 group-hover:opacity-0 rounded-md"
							src={post.images[1]}
							alt=""
						/>
					) : (
						<img
							className="relative transform duration-500 group-hover:opacity-0 rounded-md"
							src={post.images[0]}
							alt=""
						/>
					)}
				</div>
				<ul className="mt-6 font-semibold text-gray-500 text-left ml-1">
					<li className="inline mr-3 pb-1 border-b-2 border-green-500">
						{post.category.name}
					</li>
				</ul>
				<p className="mt-6 text-xl leading-relaxed text-white">{post.title}</p>

				<p className="text-gray-400 mt-2 font-semibold text-sm">
					{post.price} €
				</p>

				<div className="flex flex-row justify-between">
					<Button
						onPress={() => {
							navigate(`/wallashop/post/${post.id}`);
						}}
						className="mt-6 border border-green-600 rounded text-white bg-[#18181B] hover:bg-green-500"
					>
						Details
					</Button>

					{!(post.stillValid === null || post.stillValid === undefined) &&
						user.id !== post.user.id && (
							<div className="relative mt-6 right-1 ml-auto flex-2 text-small">
								Still Valid:
								<br />
								<span className="text-green-600">{date}</span>
								<span className="text-green-600">{" " + time}</span>
							</div>
						)}

					{user?.id === post.user.id && (
						<>
							<EditPostModal post={post} />

							<Button
								onPress={() => handleDelete()}
								className="mt-6 border border-red-600 rounded text-white bg-[#18181B] hover:bg-red-500"
							>
								<RiDeleteBin7Line size={20} />
							</Button>
						</>
					)}
				</div>
				{!(post.stillValid === null || post.stillValid === undefined) &&
					user.id === post.user.id && (
						<div className="relative mt-6 right-1 ml-auto flex-2 text-small">
							Still Valid:
							<br />
							<span className="text-green-600">{date}</span>
							<span className="text-green-600">{" " + time}</span>
						</div>
					)}
			</article>
		</>
	);
};

export default PostCard;
