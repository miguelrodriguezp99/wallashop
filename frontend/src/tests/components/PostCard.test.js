import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import PostCard from '../../modules/app/components/PostCard';
import { LogginContext } from "../../modules/app/context/loginContext"; 
import '@testing-library/jest-dom';

const fakeLogginContext = {
    token: 'testToken',
    user: { id: 'user1' }
};

const samplePost = {
  id: '1',
  type: 'OFFER',
  images: ['image1', 'image2'],
  category: { name: 'cat' },
  title: 'title',
  price: '20',
  stillValid: '2023-12-31T12:00:00.000Z',
  user: { id: 'user1' },
};


jest.mock('../../backend/postService', () => ({
    ...jest.requireActual('../../backend/postService'),
    deletePost: jest.fn().mockResolvedValue({}), 
  }));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(),
}));
  

describe('PostCard component', () => {
    
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('matches the snapshot', () => {
        const post = {
          id: '1',
          type: 'OFFER',
          images: ['image1', 'image2'],
          category: { name: 'cat1' },
          title: 'title',
          price: '20',
          user: { id: 'user1' },
          stillValid: '2023-01-01T12:00:00.000Z',
        };
    
        const { asFragment } = 
        render(
            <MemoryRouter>
                <LogginContext.Provider value={fakeLogginContext}>
                  <PostCard post={samplePost} />
                </LogginContext.Provider>
            </MemoryRouter>
        );
    
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders PostCard correctly', () => {
        render(
            <MemoryRouter>
                <LogginContext.Provider value={fakeLogginContext}>
                  <PostCard post={samplePost} />
                </LogginContext.Provider>
            </MemoryRouter>
        );

        // Check if essential elements are present
        expect(screen.getByText('title')).toBeInTheDocument();
        expect(screen.getByText('20 â‚¬')).toBeInTheDocument();
        expect(screen.getByText('Details')).toBeInTheDocument();
    });

    it('handles delete button click for user-owned post', () => {
        const deleteMock = jest.fn();
        render(
        <MemoryRouter>
            <LogginContext.Provider value={fakeLogginContext}>
                <PostCard post={samplePost} />
            </LogginContext.Provider>
        </MemoryRouter>
        );
    });

    
    it('does not render EditPostModal when user does not own the post', () => {
        const fakeLogginContextForOtherUser = {
          token: 'testToken',
          user: { id: 'user2' },
        };
    
        render(
          <MemoryRouter>
            <LogginContext.Provider value={fakeLogginContextForOtherUser}>
              <PostCard post={samplePost} />
            </LogginContext.Provider>
          </MemoryRouter>
        );
    
        expect(screen.queryByTestId('edit-post-modal')).toBeNull();
      });

      
});



