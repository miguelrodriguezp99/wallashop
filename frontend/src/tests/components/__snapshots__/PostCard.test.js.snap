// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard component matches the snapshot 1`] = `
<DocumentFragment>
  <article
    class="p-5 transform duration-300 hover:-translate-y-1 cursor-pointer group hover:shadow-white bg-cardback ml-2 rounded-md"
  >
    <div
      class="folded-ribbon text-lg bg-pink-400"
    >
      Oferta
    </div>
    <div
      class=" relative max-h-125 overflow-hidden"
    >
      <img
        alt=""
        class="absolute rounded-md"
        src="image1"
      />
      <img
        alt=""
        class="relative transform duration-500 group-hover:opacity-0 rounded-md"
        src="image2"
      />
    </div>
    <ul
      class="mt-6 font-semibold text-gray-500 text-left ml-1"
    >
      <li
        class="inline mr-3 pb-1 border-b-2 border-green-500"
      >
        cat
      </li>
    </ul>
    <p
      class="mt-6 text-xl leading-relaxed text-white"
    >
      title
    </p>
    <p
      class="text-gray-400 mt-2 font-semibold text-sm"
    >
      20 â‚¬
    </p>
    <div
      class="flex flex-row justify-between"
    >
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-unit-4 min-w-unit-20 h-unit-10 text-small gap-unit-2 [&>svg]:max-w-[theme(spacing.unit-8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover mt-6 border border-green-600 rounded text-white bg-[#18181B] hover:bg-green-500"
        type="button"
      >
        Details
      </button>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-unit-4 min-w-unit-20 h-unit-10 text-small gap-unit-2 [&>svg]:max-w-[theme(spacing.unit-8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover mt-6 border border-green-600 rounded text-white bg-[#18181B] hover:bg-green-500"
        type="button"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.7574 2.99666L14.7574 4.99666H5V18.9967H19V9.2393L21 7.2393V19.9967C21 20.5489 20.5523 20.9967 20 20.9967H4C3.44772 20.9967 3 20.5489 3 19.9967V3.99666C3 3.44438 3.44772 2.99666 4 2.99666H16.7574ZM20.4853 2.09717L21.8995 3.51138L12.7071 12.7038L11.2954 12.7062L11.2929 11.2896L20.4853 2.09717Z"
          />
        </svg>
      </button>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-unit-4 min-w-unit-20 h-unit-10 text-small gap-unit-2 [&>svg]:max-w-[theme(spacing.unit-8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover mt-6 border border-red-600 rounded text-white bg-[#18181B] hover:bg-red-500"
        type="button"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17 6H22V8H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V8H2V6H7V3C7 2.44772 7.44772 2 8 2H16C16.5523 2 17 2.44772 17 3V6ZM18 8H6V20H18V8ZM9 4V6H15V4H9Z"
          />
        </svg>
      </button>
    </div>
    <div
      class="relative mt-6 right-1 ml-auto flex-2 text-small"
    >
      Still Valid:
      <br />
      <span
        class="text-green-600"
      >
        2023-12-31
      </span>
      <span
        class="text-green-600"
      >
         12:00:00
      </span>
    </div>
  </article>
</DocumentFragment>
`;
